domain/network/state/linklayer_merge.go

```diff
@@ -0,0 +1,717 @@
+// If we locate multiple existing devices with the hardware address,
+// such as will be the case for bridged NICs, fallback through the
+// following options.
+//   - If there is a device that already has a provider ID, use that name.
+//   - If the devices are of different types, choose an ethernet device over
+//     a bridge (as observed for MAAS).
+func (st *State) normalizeLinkLayeredDevices(
+	ctx context.Context,
+	incoming []network.NetInterface,
+	devices []mergeLinkLayerDevice,
+) ([]mergeLinkLayerDevice, set.Strings, error) {
+	namelessHWAddrs := set.NewStrings()
+
+	normalizedIncoming := transform.Slice(incoming,
+		func(dev network.NetInterface) mergeLinkLayerDevice {
+			if dev.MACAddress == nil {
+				st.logger.Debugf(ctx,
+					"empty MACAddress for an incoming device")
+			}
+			return mergeLinkLayerDevice{
+				Name:       dev.Name,
+				MACAddress: deref(dev.MACAddress),
+				ProviderID: string(deref(dev.ProviderID)),
+				Type:       dev.Type,
+				Addresses: transform.Slice(dev.Addrs,
+					func(addr network.NetAddr) mergeAddress {
+						return mergeAddress{
+							Value:      addr.AddressValue,
+							ProviderID: string(deref(dev.ProviderID)),
+						}
+					}),
+			}
+		})
+
+	// Check that the incoming data is not using a provider ID for more
+	// than one device. This is not verified by transaction assertions.
+	seenProviders := set.Strings{}
+	duplicatedProviders := set.Strings{}
+	for _, dev := range normalizedIncoming {
+		if dev.ProviderID != "" {
+			continue
+		}
+		if seenProviders.Contains(dev.ProviderID) {
+			duplicatedProviders.Add(dev.ProviderID)
+		}
+		seenProviders.Add(dev.ProviderID)
+	}
+	if len(duplicatedProviders) > 0 {
+		return nil, namelessHWAddrs, errors.Errorf(
+			"unable to set provider IDs %q for multiple devices",
+			duplicatedProviders.Values(),
+		)
+	}
+
+	// If the incoming devices have names, no action is required
+	// (assuming all or none here per current known provider implementations
+	// of `NetworkInterfaces`)
+	if len(normalizedIncoming) > 0 && normalizedIncoming[0].Name != "" {
+		return normalizedIncoming, namelessHWAddrs, nil
+	}
+
+	// First get the best device per hardware address.
+	devByHWAddr := make(map[string]mergeLinkLayerDevice)
+	for _, dev := range devices {
+		hwAddr := dev.MACAddress
+
+		// If this is the first one we've seen, select it.
+		current, ok := devByHWAddr[hwAddr]
+		if !ok {
+			devByHWAddr[hwAddr] = dev
+			continue
+		}
+
+		// If we have a matching device that already has a provider ID,
+		// I.e. it was previously matched to the hardware address,
+		// make sure the same one is resolved thereafter.
+		if current.ProviderID != "" {
+			continue
+		}
+
+		// Otherwise choose a physical NIC over other device types.
+		if dev.Type == corenetwork.EthernetDevice {
+			devByHWAddr[hwAddr] = dev
+		}
+	}
+
+	// Set the names and remembers normalized nameless addresses
+	return transform.Slice(normalizedIncoming,
```

Comment:
This is confusing to read, the transform is being run only for its side effects and is not actually doing any transforming. I think this should really be a for loop for readabililty.